timelock-impl/src/main/java/com/palantir/atlasdb/timelock/lock/watch/LockEventLogImpl.java

```diff
@@ -50,58 +50,67 @@
+        if (!locksTakenOut.isEmpty()) {
+            synchronized (this) {
+                slidingWindow.add(LockEvent.builder(locksTakenOut, lockToken));
+            }
         }
-        List<LockWatchEvent> events = maybeEvents.get();
-        return LockWatchStateUpdate.success(logId, toVersion, events);
     }
 
-    private boolean isVersionStale(Optional<IdentifiedVersion> fromVersion) {
-        return !fromVersion.isPresent() || !fromVersion.get().id().equals(logId);
+    @Override
+    public void logUnlock(Set<LockDescriptor> locksUnlocked) {
+        if (!locksUnlocked.isEmpty()) {
+            synchronized (this) {
+                slidingWindow.add(UnlockEvent.builder(locksUnlocked));
+            }
+        }
     }
 
-    /**
-     * Gets a snapshot estimate, then replays all lock watch events that occurred in the meantime on top of it
-     * to correct any inconsistencies as discussed in {@link this#calculateOpenLocks(RangeSet)}.
-     */
-    private LockWatchStateUpdate attemptToCalculateSnapshot() {
-        long startVersion = slidingWindow.lastVersion();
-        SnapshotEventReplayer eventReplayer = getSnapshotEstimateAndCreateReplayer();
-        long endVersion = slidingWindow.lastVersion();
+    @Override
+    public synchronized void logLockWatchCreated(LockWatches newWatches) {
+        Set<LockDescriptor> openLocks = calculateOpenLocks(newWatches.ranges());
+        slidingWindow.add(LockWatchCreatedEvent.builder(newWatches.references(), openLocks));
+    }
 
-        Optional<List<LockWatchEvent>> eventsToReplay = slidingWindow.getFromTo(startVersion, endVersion);
-        if (!eventsToReplay.isPresent()) {
-            return LockWatchStateUpdate.failed(logId);
-        }
-        eventsToReplay.get().forEach(eventReplayer::replay);
-        return LockWatchStateUpdate.snapshot(
-                logId,
-                endVersion,
-                eventReplayer.locked,
-                eventReplayer.watches);
+    private boolean shouldCalculateSnapshot(Optional<IdentifiedVersion> fromVersion) {
+        return !fromVersion.isPresent() || !fromVersion.get().id().equals(logId) || !slidingWindow.contains(
+                fromVersion.get().version());
     }
 
-    private SnapshotEventReplayer getSnapshotEstimateAndCreateReplayer() {
+    private LockWatchStateUpdate calculateSnapshot() {
+        long lastVersion = slidingWindow.lastVersion();
         LockWatches currentWatches = watchesSupplier.get();
-        Set<LockWatchReference> watched = new HashSet<>(currentWatches.references());
-        Set<LockDescriptor> estimatedLocks = calculateOpenLocks(currentWatches.ranges());
-        return new SnapshotEventReplayer(watched, estimatedLocks);
+        Set<LockWatchReference> watches = new HashSet<>(currentWatches.references());
+        Set<LockDescriptor> openLocks = calculateOpenLocks(currentWatches.ranges());
```

Comment:
The thing to understand here is what does this snapshot actually mean: it's a best effort thing to indicate if there are any open locks and return them. This snapshot isn't really a *consistent* snapshot: while we're locked here in the log, the HeldLocksCollection is still *granting* locks, which, as soon as we unlock the log, will add their events. But as per the RFC, this doesn't matter for any transactions that end up working with this snapshot: what matters for conflict checking is that there were no events for cached rows in between the transaction starting and trying to commit.

After these changes (and possibly before them), if we compute a snapshot, the node might later receive events it already knows about from the snapshot again. And that should be fine.